nixos/modules/services/networking/frp.nix

```
@@ -0,0 +1,99 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.frp;
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "frp.ini" cfg.settings;
+  isClient = (cfg.role == "client");
+  isServer = (cfg.role == "server");
+in
+{
+  options = {
+    services.frp = {
+      enable = mkEnableOption (lib.mdDoc "");
+
+      package = mkPackageOptionMD pkgs "frp" { };
+
+      role = mkOption {
+        type = types.enum [ "server" "client" ];
+        default = "server";
+        description = lib.mdDoc ''
+          The frp consists of `client` and `server`. The server is usually
+          deployed on the machine with a public IP address, and
+          the client is usually deployed on the machine
+          where the Intranet service to be penetrated resides.
+        '';
+      };
+
+      settings = mkOption {
+        type = settingsFormat.type;
+        default = { };
+        description = lib.mdDoc ''
+          Frp configuration, for configuration options
+          see the example of [client](https://github.com/fatedier/frp/blob/dev/conf/frpc_full.ini)
+          or [server](https://github.com/fatedier/frp/blob/dev/conf/frps_full.ini) on github.
+        '';
+        example = literalExpression ''
+          {
+            common = {
+              server_addr = "x.x.x.x";
+              server_port = 7000;
+            };
+          }
+        '';
+      };
+    };
+  };
+
+  config =
+    let
+      serviceWants = lib.optionals isClient [ "network-online.target" ];
+      serviceAfter = if isClient then [ "network-online.target" ] else [ "network.target" ];
+      serviceRestrictAddressFamilies = [ "AF_INET" "AF_INET6" ] ++ lib.optionals isClient [ "AF_UNIX" ];
+      serviceCapability = lib.optionals isServer [ "CAP_NET_BIND_SERVICE" ];
+      serviceStateDirectoryMode = lib.optionalString isServer "0700";
+      serviceUMask = lib.optionalString isServer "0007";
+      executableFile = if isClient then "frpc" else "frps";
+    in
+    lib.mkIf cfg.enable {
+      systemd.services = {
+        frp = {
+          wants = serviceWants;
+          after = serviceAfter;
+          wantedBy = [ "multi-user.target" ];
+          description = "A fast reverse proxy to help you expose a local server behind a NAT or firewall to the internet.";
+          serviceConfig = {
+            Type = "simple";
+            Restart = "on-failure";
+            RestartSec = 15;
+            ExecStart = "${cfg.package}/bin/${executableFile} -c ${configFile}";
+            StateDirectoryMode = serviceStateDirectoryMode;
+            UMask = serviceUMask;
+            # Hardening
+            DynamicUser = true;
```

`DynamicUser` is not usually considered as an hardening option.
```suggestion
            DynamicUser = true;
            # Hardening
            UMask = serviceUMask;
```

---

Done.